File: utilslib/src/main/java/com/ashlikun/utils/other/IntentUtils.java
package com.ashlikun.utils.other;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.provider.MediaStore;

import androidx.fragment.app.Fragment;

import com.ashlikun.utils.AppUtils;
import com.ashlikun.utils.ui.ActivityManager;

import java.io.File;
import java.io.IOException;

/**
 * 作者　　: 李坤
 * 创建时间: 2018/9/19　9:42
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：系统一些意图的工具类
 */
public class IntentUtils {
    public static void jump(Intent intent) {
        Activity activity = ActivityManager.getForegroundActivity();
        if (activity != null) {
            activity.startActivity(intent);
        } else {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            AppUtils.getApp().startActivity(intent);
        }
    }

    /**
     * 根据手机号发送短信
     *
     * @param phone
     */
    public static void sendSmsByPhone(String phone) {
        if (StringUtils.isEmpty(phone)) {
            return;
        }
        Uri uri = Uri.parse("smsto:" + phone);
        Intent intent = new Intent(Intent.ACTION_SENDTO, uri);
        // intent.putExtra("sms_body", "");
        jump(intent);
    }

    /**
     * 调用系统分享
     */
    public static void shareToOtherApp( String title, String content, String dialogTitle) {
        Intent intentItem = new Intent(Intent.ACTION_SEND);
        intentItem.setType("text/plain");
        intentItem.putExtra(Intent.EXTRA_SUBJECT, title);
        intentItem.putExtra(Intent.EXTRA_TEXT, content);
        jump(Intent.createChooser(intentItem, dialogTitle));
    }

    /**
     * 回到桌面
     */
    public static void startHomeActivity() {
        Intent intent = new Intent(Intent.ACTION_MAIN);
        intent.addCategory(Intent.CATEGORY_HOME);
        jump(intent);
    }


    /**
     * 根据手机好拨打电话
     */
    public static void callPhone(String phone) {
        if (StringUtils.isEmpty(phone)) {
            return;
        }
        Intent intent = new Intent("android.intent.action.CALL", Uri.parse("tel:" + phone));
        jump(intent);
    }

    /**
     * 方法功能：开始拍照
     *
     * @param context
     * @param appSDCachePath 文件存放路径
     * @param requestCode    请求的Code
     * @return 文件，异步的
     */
    public static String startPicture(Activity context, String appSDCachePath, int requestCode) {
        File fDir = new File(appSDCachePath);
        String cachePath = System.currentTimeMillis() + ".jpg";
        if (fDir.exists() || fDir.mkdirs()) {
            File cameraFile = new File(fDir, cachePath);
            if (!cameraFile.exists()) {
                try {
                    cameraFile.createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
            intent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(cameraFile));
            context.startActivityForResult(intent, requestCode);
        }
        return cachePath;
    }

    /**
     * 方法功能：开启文件选择
     *
     * @param context
     * @param type    媒体类型 audio/*  选择音频
     *                video/*  选择视频
     *                video/*;image/*  同时选择视频和图片
     *                * /*  全部  没有空格
     */
    public static void startFileSelect(Activity context, String type, int requestCode) {
        Intent intent = new Intent();
        intent.setType(type);
        intent.setAction(Intent.ACTION_GET_CONTENT);
        intent.putExtra("return-data", true);
        context.startActivityForResult(intent, requestCode);
    }

    /**
     * 方法功能：同上
     */

    public static void startFileSelect(Fragment context, String Type, int requestCode) {
        Intent intent = new Intent();
        intent.setType(Type);
        intent.setAction(Intent.ACTION_GET_CONTENT);
        intent.putExtra("return-data", true);
        context.startActivityForResult(intent, requestCode);
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/other/IntentUtils.kt
package com.ashlikun.utils.other

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.provider.MediaStore
import android.provider.Settings
import androidx.fragment.app.Fragment
import com.ashlikun.utils.AppUtils
import com.ashlikun.utils.AppUtils.app
import com.ashlikun.utils.ui.ActivityManager
import com.ashlikun.utils.ui.fActivity
import java.io.File
import java.io.IOException

/**
 * @author　　: 李坤
 * 创建时间: 2021.12.13 16:06
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：系统一些意图的工具类
 */
inline fun Intent.jump() = IntentUtils.jump(this)

object IntentUtils {
    fun jump(intent: Intent) = runCatching {
        if (fActivity != null) {
            fActivity?.startActivity(intent)
        } else {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            app.startActivity(intent)
        }
    }.onFailure { it.printStackTrace() }.isSuccess

    /**
     * 根据手机号发送短信
     *
     * @param phone
     */
    fun sendSmsByPhone(phone: String): Boolean {
        if (phone.isEmpty()) {
            return false
        }
        val uri = Uri.parse("smsto:$phone")
        val intent = Intent(Intent.ACTION_SENDTO, uri)
        // intent.putExtra("sms_body", "");
        return jump(intent)
    }

    /**
     * 调用系统分享
     */
    fun shareToOtherApp(title: String, content: String, dialogTitle: String): Boolean {
        val intentItem = Intent(Intent.ACTION_SEND)
        intentItem.type = "text/plain"
        intentItem.putExtra(Intent.EXTRA_SUBJECT, title)
        intentItem.putExtra(Intent.EXTRA_TEXT, content)
        return jump(Intent.createChooser(intentItem, dialogTitle))
    }

    /**
     * 回到桌面
     */
    fun startHomeActivity(): Boolean {
        val intent = Intent(Intent.ACTION_MAIN)
        intent.addCategory(Intent.CATEGORY_HOME)
        return jump(intent)
    }

    /**
     * 根据手机好拨打电话
     */
    fun callPhone(phone: String): Boolean {
        if (phone.isEmpty()) {
            return false
        }
        return jump(Intent("android.intent.action.CALL", Uri.parse("tel:$phone")))
    }

    /**
     * 开始拍照
     *
     * @param context
     * @param appSDCachePath 文件存放路径
     * @param requestCode    请求的Code
     * @return 文件，异步的
     */
    fun startPicture(context: Activity, appSDCachePath: String, requestCode: Int): String {
        val fDir = File(appSDCachePath)
        val cachePath = System.currentTimeMillis().toString() + ".jpg"
        if (fDir.exists() || fDir.mkdirs()) {
            val cameraFile = File(fDir, cachePath)
            if (!cameraFile.exists()) {
                try {
                    cameraFile.createNewFile()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            intent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(cameraFile))
            context.startActivityForResult(intent, requestCode)
        }
        return cachePath
    }

    /**
     * 开启文件选择
     *
     * @param context
     * @param type    媒体类型 audio/ *  选择音频
     * video/ *  选择视频
     * video/ *;image/ *  同时选择视频和图片
     * * / *  全部  没有空格
     */
    fun startFileSelect(context: Activity, type: String, requestCode: Int) {
        val intent = Intent()
        intent.type = type
        intent.action = Intent.ACTION_GET_CONTENT
        intent.putExtra("return-data", true)
        context.startActivityForResult(intent, requestCode)
    }

    /**
     * 同上
     */
    fun startFileSelect(context: Fragment, type: String, requestCode: Int) {
        val intent = Intent()
        intent.type = type
        intent.action = Intent.ACTION_GET_CONTENT
        intent.putExtra("return-data", true)
        context.startActivityForResult(intent, requestCode)
    }

    /**
     * 应用详细页
     */
    fun startMyAppSetting() = jump(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
        data = Uri.fromParts("package", AppUtils.packageName, null)
    })

    /**
     * 系统设置界面
     */
    fun startSysSetting() = jump(Intent(Settings.ACTION_SETTINGS))
}